Expose commit memory used in WindowsMemoryUsage struct

kiashok opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds a field to track the commit memory usage bytes to container stats. 
In Linux, a process is charged with memory usage only based on the total amount of physical memory (working set bytes) it uses. However, on Windows, the total physical and virtual memory (commit memory) currently in use by the process, is used to account for its total memory consumption.
Because of this, auto scaling windows containers etc should rely on the commit memory used rather than the working set memory. HCS surfaces the commit memory used to hcsshim but it isn't getting tracked at the CRI layer. This PR will help to expose the total commit memory usage as part of container stats structure.

### NOTE: The previous PR closed unexpectedly while rebasing to resolve conflicts. (Old PR - https://github.com/kubernetes/kubernetes/pull/117127) 

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

No

```release-note
CRI: expose commit memory bytes in container stats specific to Windows
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig node
/sig windows

/cc @jsturtevant @marosset

kiashok committed:
Expose commit memory used in WindowsMemoryUsage struct

Signed-off-by: kiashok <kiashok@microsoft.com>

kiashok commented:
/cc @SergeyKanzhelev @pacoxu @furkatgofurov7

furkatgofurov7 commented:
/ok-to-test

jsturtevant commented:
/triage accepted
/priority important-soon

jsturtevant commented:
/lgtm

SergeyKanzhelev commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1694,6 +1694,8 @@ message WindowsMemoryUsage {
     UInt64Value available_bytes = 3;
     // Cumulative number of page faults.
     UInt64Value page_faults = 4;
+    // Total commit memory in use. Commit memory is total of physical and virtual memory in use.
```
I still don't understand why wouldn't it be best to change comment as suggested here: https://github.com/kubernetes/kubernetes/pull/117127#discussion_r1244486342 in this same commit. Changing implementation while comment stays to be incorrect doesn't seem to be a good approach. Sorry if I am missing something

kiashok replied:
We are NOT changing any behavior with adding this field. hcsshim does not even expose or return this commit memory. I feel adding and changing the comment of available memory even before we expose the commit memory used is incorrect. We it would be good to change it only once we actually populate and expose this field from the shim.

kiashok replied:
please refer to https://github.com/kubernetes/kubernetes/pull/117127#discussion_r1228816818
https://github.com/kubernetes/kubernetes/pull/117127#discussion_r1229730904

pacoxu replied:
> We are NOT changing any behavior with adding this field. hcsshim does not even expose or return this commit memory. I feel adding and changing the comment of available memory even before we expose the commit memory used is incorrect. We it would be good to change it only once we actually populate and expose this field from the shim.

I would +1 for @SergeyKanzhelev as CRI definition is a guide for implementations.
- If we plan to do that which is designed or in scheduling, we should let others know the CRI design in this spec.

kubelet can implement things based on the design using commit memory before the implementation in container runtimes and other parts. 

> I do think we will want to update available_bytes to be limit - committed_memory once commit memory is plumbed through all the layers. from https://github.com/kubernetes/kubernetes/pull/117127#discussion_r1228816818.

kiashok replied:
There has been quite a bit of back and forth for the past two months on this - I'll update the comment now and update the PR.
However, IMO I would preferred to have updated the code comment only after the values are plumbed through as the behavior till then for available bytes is still going to max - working set memory and might cause more confusion if updated now since there this variable is not even getting populated today.

pacoxu replied:
If there are some special changes here， that should be mentioned as well if necessary.
- comments is intented to make things clearer， not into a mess.

SergeyKanzhelev replied:
You can change the comment into the new behavior and add a note after the intended behavior that the semantic of this field was different before.

As @pacoxu said CRI interface express the intended bnehavior and should be changed first

pacoxu commented:
An update to the CRI change log in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cri-api/README.md#v127 would be appreciated for v1.28.

kiashok committed:
Update cri-api change in v1.28.0

Signed-off-by: kiashok <kiashok@microsoft.com>

SergeyKanzhelev reviewed (approved):
/lgtm

thank you for addressing comments and extra thanks for adding a changelog

SergeyKanzhelev commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1690,10 +1690,12 @@ message WindowsMemoryUsage {
     // The amount of working set memory in bytes.
     UInt64Value working_set_bytes = 2;
-    // Available memory for use. This is defined as the memory limit - workingSetBytes.
+    // Available memory for use. This is defined as the memory limit - commit_memory_bytes.
```
if you want to add the additional note about past behavior, please do it. I lgtm either way, just unhold of you do not plan to add a note

/hold

kiashok replied:
I have the changes for exposing this field from hcsshim and containerd already. Once this is checked in and tag for 1.28 is cut, I'll send a PR for them immediately.

/unhold

kiashok commented:
> An update to the CRI change log in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cri-api/README.md#v127 would be appreciated for v1.28.

@pacoxu I have updated the change log. Please let me know if things look ok! Thanks.

SergeyKanzhelev reviewed (approved):
re-approving to recalculate approvals. Should be ok now

kiashok commented:
> re-approving to recalculate approvals. Should be ok now

All the tests are passing. Could we merge the PR if everything looks ok?

kiashok commented:
@SergeyKanzhelev @pacoxu Looks like another PR was merged 8 hours back that touches the proto files that this PR did causing some conflicts. The change log also needs to be modified to include this new PR. Working on these now.

kiashok commented:
> @SergeyKanzhelev @pacoxu Looks like another PR was merged 8 hours back that touches the proto files that this PR did causing some conflicts. The change log also needs to be modified to include this new PR. Working on these now.

@SergeyKanzhelev  @pacoxu @markllama  @jsturtevant  Pushed new changes to resolve the conflict and add the new PR that was merged this morning. LGTMs have been reset. Could you please take a look?

pacoxu commented:
/lgtm

pacoxu commented:
need rebase again

kiashok commented:
> need rebase again

Rebased again and updated the config log to include another change that was merged last night! Could you please take a look? @pacoxu @SergeyKanzhelev @marosset @jsturtevant

SergeyKanzhelev reviewed (approved):
/lgtm
